### PR TITLE
axes scales removed

### PR DESCRIPTION
**Pull request by abuts**
The work fully modified the binning algorithm to avoid axes_scales method which make it more robust and applicable for generic projections. Further improvements are possible but would be the subject of other tickets. 

I think, this ticket brings the state of generic projection epic to satisfactory conditions

### CODE REVIEW

File: _test/test_transformation/test_axes_block_interpolation.m [hunk 1]
```diff
@@ -143,16 +137,16 @@ function test_interp_1D_peak_averaging(~)
...
             ab_interp = ortho_axes('img_range',dbr,'nbins_all_dims',[8,1,1,1]);
 
-            si = ab_interp.interpolate_data(int_points,int_data,cell_sizes);
+            proj = ortho_proj();
+            si = ab_interp.interpolate_data(ab_base,proj ,data,proj );
 
             assertEqualToTol(sum(si),sum(data),1.e-12);
-            sample= [0,0,1.25,7.5,1.25,0,0,0];
-            assertElementsAlmostEqual(si,sample')
+            %sample= [0,0,1.25,7.5,1.25,0,0,0];
```
Comment (oerc0122) [inline _test/test_transformation/test_axes_block_interpolation.m:148]: Again

File: horace_core/sqw/@DnDBase/targ_range.m [hunk 2]
```diff
@@ -59,7 +59,7 @@
 [range0,dE_range] = transf_range(obj,source_proj,1);
 difr = 1;
 node_mult = 2;
-while difr>1.e-3 && node_mult<17
+while difr>1.e-3 && node_mult<33 % 2,4,8,16,32 % it is hull, so acceptable number of points
```
Comment (oerc0122) [inline horace_core/sqw/@DnDBase/targ_range.m:62]: Why? Must be factor of two? What is acceptable?

File: horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m [hunk 3]
```diff
@@ -319,42 +319,31 @@
...
             % radius
-            [~,step] = obj.get_axes_scales();
-            volume = prod(step);
+            if nargin == 1
+                [~,~,~,volume] = obj.get_bin_nodes('-axes_only');                
+                return;
+            else
+                ax = varargin{1};
+            end
+            volume = obj.calc_bin_volume(ax);
```
Comment (oerc0122) [inline horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m:346]:
```suggestion
            if isempty(varargin)
                [~,~,~,volume] = obj.get_bin_nodes('-axes_only');
            elseif numel(varargin) == 1
                ax = varargin{1};
                volume = obj.calc_bin_volume(ax);
            else
                error('HORACE:get_bin_volumes:invalid_argument', ...
                    'Only one argument should be passed, received %d', numel(varargin))
            end
```

File: horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m [hunk 4]
```diff
@@ -371,8 +360,8 @@
             ax_block_al = realign_bin_edges_(obj,ax_block);
         end
         %
-        function [interp_points,density,cell_sizes] = get_density(obj,datasets)
-            % Convert input datasets defined on centerpoints of this grid
+        function [interp_points,density] = get_density(obj,datasets)
+            % Convert input datasets defined on center-points of this grid
```
Comment (oerc0122) [inline horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m:364]:
```suggestion
            % Convert input datasets defined on centre-points of this grid
```

File: horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m [hunk 5]
```diff
@@ -386,40 +375,43 @@
...
             % density
             %          -- cellarray of density points calculated in the
             %             density points positions.
             %             Number of cells in the output array is equal to
             %             the number of input datasets
-            % cell_sizes
-            %          -- 4-elements vector containing the sizes of the
-            %             cell (to be extended on heterogheneous cell)
+            % bin_volume
+            %          -- the volume of each density cell. Single value if
```
Comment (oerc0122) [inline horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m:385]:
```suggestion
            %          -- the volume of each density cell. Scalar if
```

File: _test/test_dnd_class/test_dnd_cut.m [hunk 6]
```diff
@@ -3,52 +3,66 @@
...
+        % tests
+        function test_2D_to2D_cut_with_spher_proj(obj)
+            ws = warning('off','HORACE:runtime_error');
+            clOb = onCleanup(@()warning(ws));
+            
+            proj = spher_proj();
+            cut_range = obj.d2d_obj.targ_range(proj);
+            w2 = cut(obj.d2d_obj,proj,[1,0.1,2], ...
+                [113,114],[cut_range(1,3),0.1,cut_range(2,3)],[-0.25,0.25]);
+            skipTest('#983 does this spherical projection work correctly?')
```
Comment (cmarooney-stfc) [inline _test/test_dnd_class/test_dnd_cut.m:39]: if this works formally we should not be skipping this test, we should be putting in a ticket for the scientific validation

File: horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m [hunk 7]
```diff
@@ -433,15 +425,10 @@
             if nargin < 5
```
Comment (oerc0122) [inline horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m:425]:
```suggestion
            if isempty(varargin)
```

File: horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_nodes_.m [hunk 8]
```diff
@@ -1,20 +1,23 @@
...
 %              axes points.
 % halo      -- if true, build one-cell width halo around the generated axes
 %              grid. Not building halo along energy axes in 3D mode
-% bin_edges
-%           -- if true, return grid used to define density, namely with points
-%              located on the grid cell edges + edges of integrated
-%              dimensions.
+% bin_edges -- if true, return grid containing bin edges
+% dens_interp
+%           -- if true, return grid used to define density, bin centers for
```
Comment (oerc0122) [inline horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_nodes_.m:19]:
```suggestion
%           -- if true, return grid used to define density, bin centres for
```

File: _test/test_transformation/test_AxesBlockBase_properties.m [hunk 9]
```diff
@@ -31,6 +31,37 @@
...
+            ax{3} = 1:0.1:2;
+            ax{4} = [0,2,3,4,5,10];
+            n_cells = cellfun(@(x)numel(x)-1,ax);
+            n_cells = prod(n_cells);
+
+            bv = ab.get_bin_volume(ax);
+
+            assertEqual(numel(bv),n_cells);
+
+            assertEqualToTol(bv(1),1*1*0.1*2,'tol',1.e-11);
```
Comment (cmarooney-stfc) [inline _test/test_transformation/test_AxesBlockBase_properties.m:50]: worth a comment on why these numbers in the comparison are correct - i.e. where they came from

File: horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_volume_.m [hunk 10]
```diff
@@ -0,0 +1,34 @@
+function volume = calc_bin_volume_(~,ax_in_cell)
```
Comment (oerc0122) [inline horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_volume_.m:1]: If this doesn't take `obj` should be static

File: _test/test_transformation/test_axes_block_interpolation.m [hunk 11]
```diff
@@ -112,20 +106,20 @@ function test_interp_1D_frac_points_int_coeff(~)
             ax = ab_base.p{1};
             cp = 0.5*(ax(1:end-1)+ax(2:end));
             data = ones(size(cp));
-            [int_points,int_data,cell_sizes] = ab_base.get_density(data);
 
-            % define bins to give exactly the same range as for ab_base
+            % define less bins buy the same range as for ab_base
```
Comment (cmarooney-stfc) [inline _test/test_transformation/test_axes_block_interpolation.m:110]:
```suggestion
            % define less bins by the same range as for ab_base
```
I understood the old comment, but I don't understand this one.

File: _test/test_transformation/test_axes_block_interpolation.m (hunk 1 above)
Comment (cmarooney-stfc) [inline _test/test_transformation/test_axes_block_interpolation.m:148]: can we get rid of this line?

File: _test/test_transformation/test_cut_parameters.m [hunk 12]
```diff
@@ -47,7 +47,7 @@ function test_default_spher_from_ortho(obj)
             assertElementsAlmostEqual(targ_range(:,1)',[0,sqrt(1.5^2+0.5^2)], ...
                 'absolute',1.e-2)
             assertEqual(targ_range(:,2)',[0,90])
-            assertElementsAlmostEqual(targ_range(:,3)',[-179.9642,180], ...
+            assertElementsAlmostEqual(targ_range(:,3)',[-179.9821,180], ...
```
Comment (cmarooney-stfc) [inline _test/test_transformation/test_cut_parameters.m:50]: worth commenting why we need this much precision of this constant

File: horace_core/sqw/coord_transform/@aProjectionBase/aProjectionBase.m [hunk 13]
```diff
@@ -311,8 +277,100 @@
...
+            else
+                if obj.do_3D_transformation
+                    if any(source_ax_block.pax == 4)
+                        ax_num = 1:4;
+                    else
+                        ax_num = 1:3;
+                    end
+                else
+                    ax_num = 1:4;
+                end
```
Comment (oerc0122) [inline horace_core/sqw/coord_transform/@aProjectionBase/aProjectionBase.m:371]:
```suggestion
                if obj.do_3D_transformation && ~any(source_ax_block.pax == 4)
                    ax_num = 1:3;
                else
                    ax_num = 1:4;
                end
```

File: _test/test_transformation/test_ortho_axes.m [hunk 14]
```diff
@@ -625,7 +625,11 @@ function test_get_bin_centers_2D_4d(~)
                 [dbr(1,3),dbr(2,3)];[dbr(1,4),1,dbr(2,4)]};
             ab = ortho_axes(bin0{:});
 
-            [nodes,en,nbins] = ab.get_bin_nodes('-bin_centre',(dbr(2,:)-dbr(1,:))'/10);
+            char_size_des = (dbr(2,:)-dbr(1,:))/10;
+            char_size_ex  = (ab.img_range(2,:)-ab.img_range(1,:))./ab.nbins_all_dims;
+            mult = ceil(char_size_ex./char_size_des);
+            mult(mult<1) = 1;
+            [nodes,en,nbins] = ab.get_bin_nodes('-bin_centre',mult);
```
Comment (cmarooney-stfc) [inline _test/test_transformation/test_ortho_axes.m:632]: this code from L.628 is highly opaque. Can you provide a description of what it is, in particular how it improves on the old dbr-dr/10

**Review by oerc0122 — CHANGES_REQUESTED** [review]
Could do with some cleaning up

File: _test/test_transformation/test_ortho_axes.m [hunk 15]
```diff
@@ -654,24 +658,31 @@ function test_get_bin_centers_2D_native(~)
             assertEqual(size(centers,2),the_size);
         end
         %
-        function test_worong_char_size_throw(~)
+        function test_worong_mult_throw(~)
```
Comment (cmarooney-stfc) [inline _test/test_transformation/test_ortho_axes.m:661]:
```suggestion
        function test_wrong_mult_throw(~)
```

File: _test/test_transformation/test_ortho_axes.m [hunk 16]
```diff
@@ -654,24 +658,31 @@ function test_get_bin_centers_2D_native(~)
...
             dbr = [-1,-2,-3,0;1,2,3,10];
             bin0 = {[dbr(1,1),0.1,dbr(2,1)];[dbr(1,2),dbr(2,2)];...
                 [dbr(1,3),dbr(2,3)];[dbr(1,4),1,dbr(2,4)]};
             ab = ortho_axes(bin0{:});
-            ex = assertExceptionThrown(@()get_bin_nodes(ab,'-bin_centre',(dbr(2,:)-dbr(1,:))/10),...
+            mult = ones(3,1);
+            mult(mult<1) = 1;
```
Comment (cmarooney-stfc) [inline _test/test_transformation/test_ortho_axes.m:667]: as mult from L.666 is all ones, this does not seem to do anything. What are you trying to do here?

File: _test/test_transformation/test_ortho_axes.m [hunk 17]
```diff
@@ -654,24 +658,31 @@ function test_get_bin_centers_2D_native(~)
...
+            ex = assertExceptionThrown(@()get_bin_nodes(ab,'-bin_centre',mult),...
                 'HORACE:AxesBlockBase:invalid_argument');
-            assertTrue(strncmp(ex.message,'characteristic size, if present',31));
+            assertTrue(strncmp(ex.message,'nnodes multipler should',23));
         end
         %
         function test_worong_keyword_throw(~)
```
Comment (cmarooney-stfc) [inline _test/test_transformation/test_ortho_axes.m:673]:
```suggestion
        function test_wrong_keyword_throw(~)
```

File: _test/test_transformation/test_ortho_axes.m [hunk 18]
```diff
@@ -654,24 +658,31 @@ function test_get_bin_centers_2D_native(~)
...
             dbr = [-1,-2,-3,0;1,2,3,10];
             bin0 = {[dbr(1,1),0.1,dbr(2,1)];[dbr(1,2),dbr(2,2)];...
                 [dbr(1,3),dbr(2,3)];[dbr(1,4),1,dbr(2,4)]};
             ab = ortho_axes(bin0{:});
-            ex = assertExceptionThrown(@()get_bin_nodes(ab,'-wrong',(dbr(2,:)-dbr(1,:))'/10),...
+            char_size_des = (dbr(2,:)-dbr(1,:))'/10;
+            char_size_ex  = (dbr(2,:)-dbr(1,:))'./ab.nbins_all_dims;
+            mult = ceil(char_size_ex./char_size_des);
+            mult(mult<1) = 1;
```
Comment (cmarooney-stfc) [inline _test/test_transformation/test_ortho_axes.m:681]: also explanation for this

File: _test/test_transformation/test_spher_axes.m [hunk 19]
```diff
@@ -17,6 +17,38 @@
...
+        function test_bin_voulume_array(~)
+            dbr = ...
+                [0.5, 45,  0, 0;...
+                1.5 , 46, 90, 10];
+            ab = spher_axes('img_range',dbr,'nbins_all_dims',[10,10,10,10]);
+
+            vol = ab.get_bin_volume();
+            assertEqual(numel(vol),10^4);
+        end       
+        function test_bin_voulume_single(~)
```
Comment (cmarooney-stfc) [inline _test/test_transformation/test_spher_axes.m:29]:
```suggestion
        function test_bin_volume_single(~)
```

File: horace_core/sqw/@DnDBase/private/cut_single_.m [hunk 20]
```diff
@@ -34,10 +34,12 @@
...
 %proj_given = true;
 if proj_given
-    % Interpolate image and accumulate interpolated data for cut
+    % Interpolate image on non-commensurate grid and accumulate interpolated 
+    % data for cut
     [s, e, npix] = cut_interpolate_data_( ...
-        w, tag_proj,targ_axes);%
+        w, tag_proj,targ_axes);
 else
+    % ingegrate signal and error withing commensurate grids
```
Comment (cmarooney-stfc) [inline horace_core/sqw/@DnDBase/private/cut_single_.m:42]:
```suggestion
    % integrate signal and error within commensurate grids
```

File: horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m [hunk 21]
```diff
@@ -386,40 +375,43 @@
...
             %          -- cellarray of density points calculated in the
             %             density points positions.
             %             Number of cells in the output array is equal to
             %             the number of input datasets
-            % cell_sizes
-            %          -- 4-elements vector containing the sizes of the
-            %             cell (to be extended on heterogheneous cell)
+            % bin_volume
+            %          -- the volume of each density cell. Single value if
+            %             cells have the same size or numel(density) if
```
Comment (cmarooney-stfc) [inline horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m:386]: Last if is dangling. Do you need it?

File: horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_volume_.m [hunk 22]
```diff
@@ -0,0 +1,34 @@
...
+%CALC_BIN_VOLUME_ calculate the volume of a lattice cell defined by the
+%cellarray of lattice axes.
```
Comment (cmarooney-stfc) [inline horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_volume_.m:3]: is it worth saying what the units are? Particularly if we are looking at hklE so the axes are heterogeneous

File: _test/test_dnd_class/test_dnd_cut.m [hunk 23]
```diff
@@ -3,52 +3,66 @@
...
 
-            obj = obj@TestCaseWithSave(argi{:});            
+            obj = obj@TestCaseWithSave(argi{:});
             hp = horace_paths();
             dnd_2d_fullpath = fullfile(hp.test_common,obj.dnd_file_2d_name);
-            %hor_root = horace_root();            %            
+            %hor_root = horace_root();            %
             %dnd_2d_fullpath = fullfile(hor_root,'_test/common_data',...
-            %    obj.dnd_file_2d_name);            
+            %    obj.dnd_file_2d_name);
```
[the remark this answers is not in the document]
Comment (abuts) [reply _test/test_dnd_class/test_dnd_cut.m:25]: yes, removed

File: _test/test_transformation/test_axes_block_interpolation.m (hunk 1 above)
Comment (abuts) [reply _test/test_transformation/test_axes_block_interpolation.m:148]: this one I would leave. Despite current algorithm gives better accuracy then previous one, I was sure that the previous was correct too.  Do not have time to investigae

File: _test/test_dnd_class/test_dnd_cut.m (hunk 6 above)
Comment (abuts) [reply _test/test_dnd_class/test_dnd_cut.m:39]: this what it is done. I wrote the ticket, all working parts of the test are running but the result is questionable.

File: horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_volume_.m (hunk 10 above)
Comment (abuts) [reply horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_volume_.m:1]: spherical projection overload do take object

File: _test/test_transformation/test_ortho_axes.m (hunk 16 above)
Comment (abuts) [reply _test/test_transformation/test_ortho_axes.m:667]: copied from other place. Fixed

File: horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m (hunk 21 above)
Comment (abuts) [reply horace_core/sqw/axes_blocks/@AxesBlockBase/AxesBlockBase.m:386]: Removed

File: horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_volume_.m (hunk 22 above)
Comment (abuts) [reply horace_core/sqw/axes_blocks/@AxesBlockBase/private/calc_bin_volume_.m:3]: no, whatever axes are, the units are. The axes for hkldE still homogeneous.

**Pull request merged**